Add support for VM data sets

dekp opened the pull request:
## Description

Implement support for the [VM data sets API](https://developer.vmware.com/apis/vsphere-automation/latest/vcenter/vm/data_sets/) introduced in vSphere 8.0. 

[Data sets](https://core.vmware.com/resource/vsphere-datasets) provide a mechanism to share data between the vSphere management layer (i.e. vCenter and ESX) and the VM guest OS. See the [VMware Guest SDK Programming Guide](https://developer.vmware.com/docs/16962/vmware-guest-sdk-programming-guide/GUID-AEC8BE71-FA22-4A87-8C6F-F4EBE95AA8DE.html) for details on using data sets from within the guest OS of a VM.

### Guest support for data sets
Using data sets from within a VM guest OS seems to be out of scope for `govmomi` at the moment:
`govmomi` has a guest `toolbox` sub-project but it is an alternative to the VMware tools daemon. `govmomi` does not have an alternative to the VMware tools CLI (i.e. `vmtoolsd --cmd`), nor to the VMware [Guest SDK](https://developer.vmware.com/docs/16962/vmware-guest-sdk-programming-guide/GUID-AEC8BE71-FA22-4A87-8C6F-F4EBE95AA8DE.html).
Data sets are stored on the ESXi host. The data sets feature does not need any collaboration by the VMware tools daemon to function, so there is no need to implement anything in the guest `toolbox`.
Currently data sets can be accessed from within the VM guest OS via:
- the CLI of VMware tools, i.e. `vmtoolsd --cmd`, as described in the [DataSets blog](https://core.vmware.com/resource/vsphere-datasets)
- the VMware [Guest SDK](https://developer.vmware.com/docs/16962/vmware-guest-sdk-programming-guide/GUID-AEC8BE71-FA22-4A87-8C6F-F4EBE95AA8DE.html)

### Implementation of data sets API in vcsim
Store data sets as a field in `simulator.VirtualMachine`, because a data set can be copied during a VM clone or snapshot (see `dataset.CreateSpec.OmitFromSnapshotAndClone`). Clone and snapshot are implemented in `simulator.VirtualMachine`, so it seems more natural to keep data sets in `simulator.VirtualMachine` and access them from the code in the vapi simulator.

Extract the simulation code for REST APIs under `/api/vcenter/vm` into its own sub-simulator:
Data set request paths look like this: `/api/vcenter/vm/{moid}/data-sets/{id}`. Currently vcsim routes API requests via the standard [http.ServeMux](https://pkg.go.dev/net/http#ServeMux). However, [http.ServeMux](https://pkg.go.dev/net/http#ServeMux) does not support placeholders in the middle of the URL path (e.g. VM moid, data set id). So created a separate sub-simulator for everything under `/api/vcenter/vm`. At the moment it contains only delete VM (see https://github.com/vmware/govmomi/pull/1770) and the data sets API.

Inside the new sub-simulator for `/api/vcenter/vm`, API requests are routed via analyzing the URL path segment by segment:
As mentioned above, [http.ServeMux](https://pkg.go.dev/net/http#ServeMux) cannot handle path placeholders. Here I have done routing in code. An alternative would be to use a third-party library for request routing like [go-chi](https://github.com/go-chi/chi) or even a home-grown regex table or a radix tree. Potentially the main vapi simulator could allow sub-simulators to register for paths with placeholders. See also https://github.com/vmware/govmomi/blob/e7aac6ad473846f6a1f5bf10c524ac5a8f93022d/vapi/simulator/simulator.go#L2084

In order to get a lock on a `simulator.VirtualMachine` in the vapi simulator, had to create a `simulator.Context` similar to [SpoofContext](https://github.com/vmware/govmomi/blob/1eeb6f1c33995d1015106ea17498615ba7ad01b4/simulator/model.go#L395) but using the `simulator.Registry` received during registration.

Closes: #3050

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to
  not work as expected)
- [ ] This change requires a documentation update
- [ ] Build related change

## How Has This Been Tested?

- [x] Added `bats` tests which exercise the data sets functionality in `vcsim` and `govc`.
- [x] Developer testing of data sets support in `govc` against a real vCenter.
- [x] Developer testing of data sets support in `vcsim` via curl.

## Checklist:

- [x] My code follows the `CONTRIBUTION` [guidelines] of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged

[guidelines]: https://github.com/vmware/govmomi/blob/main/CONTRIBUTING.md

dougm reviewed (dismissed):
Awesome @dekp !  Thank you for the simulator effort in particular, I was going to approve after just reading the bats tests. Rest of the doc, code and govc command structure looks great too.

dekp commented on govc/vm/dataset/create.go:
```diff
@@ -0,0 +1,123 @@
+		return err
+	}
+	if vm == nil {
+		return errors.New("please specify a vm")
```
Sure, makes sense.

dekp commented:
Thanks for the review, @dougm !
Updated the `govc` code to return `flag.ErrHelp` if `vm` or `dataset` argument is missing. Also rebased onto the latest govc changes.

dougm reviewed (approved):
Thanks again @dekp !
